### DISCUSSION TITLE
How to generate `index.js` as ESM?

### DISCUSSION THREAD

**Discussion by raphaelmenges** [napi-rs/napi-rs#1361] [Q&A]
I am working in a ESM environment and would like to have all dependencies as module, too. However, napi-rs generates a `index.js` that follows the commonJS style, even when I define `"type": "module"` in the `package.json` [1]. I was looking through the issues and discussions but could not find any indication how to achieve this.

[1] https://www.typescriptlang.org/docs/handbook/esm-node.html

**Comment by Brooooooklyn** [comment]
The generated commonjs file compatible with both commonjs & esm modules.

For example, these imports are both valid: `import { someNativeFunction } from 'your-napi-rs-module'` or `const { someNativeFunction } = require('your-napi-rs-module')`

**Reply by raphaelmenges** [reply]
Hello @Brooooooklyn! Indeed, I manage to import the module into an ESM environment. However, I am using Vite to bundle my application the and CommonJS plugin from Rollup fails to cope with the napi-rs module:

```
[commonjs--resolver] Unexpected character '�' (Note that you need plugins to import files that are not JavaScript)
file: /Users/<app>/node_modules/<napi-rs-module>/index.js:1:0
```

This would be no issue if `napi-rs` outputs the `index.js` in ESM style, i guess.

**Reply by Brooooooklyn** [reply]
> Unexpected character '�' (Note that you need plugins to import files that are not JavaScript)

This is because your bundler is treating the `.node` file as a JavaScript file and trying to parse it.

**Reply by raphaelmenges** [reply]
Valid guess, I will investigate!

**Reply by raphaelmenges** [reply]
I have been trying to adapt the [vite configuration](https://vitejs.dev/config/build-options.html#build-commonjsoptions) such that the bundler does not attempt to parse the `.node`:

```
commonjsOptions: {
    exclude: ["*.node"],
},
```

However, the error persists. I assume this is rather an issue of rollup than napi-rs! 😕

**Reply by raphaelmenges** [reply]
I managed to make it work, adding `ssr: true` to my vite configuration, like done here: <https://github.com/cawa-93/vite-electron-builder/blob/4a788e5bea335155d9b1df86591fdd40097df950/packages/main/vite.config.js#L22>

I assume making vite treat the package like "server-side-rendering", native modules are better supported?
